File: src/glol.ml
(* Copyright (c) 2012 Ashima Arts. All rights reserved.
 * Author: David Sheets
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 *)

(* Let's link some glo! *)
open Glo_lib

(* We'll need some maps from strings to structures. *)
module M = Map.Make(String)

(* Each unit of SL source belongs to a labeled glo and has a (mostly)
   meaningless glo index that determines link precedence. *)
type unit_addr = string * int

(* To satisfy the various symbol dependency constraints, we zip through the
   link list tracking Required symbols, Top symbols satisfied, and Bottom symbols
   exposed by the above teeth. *)
type tooth = { rsym : string list;   rmac : string list;
               tsym : unit_addr M.t; tmac : unit_addr M.t;
               bsym : unit_addr M.t; bmac : unit_addr M.t;
               addr : unit_addr }
    (* Bottom * Top *)
type zipper = tooth list * tooth list

exception MissingSymbol of unit_addr * string
exception MissingMacro of unit_addr * string
exception CircularDependency of unit_addr list
exception SymbolConflict of string * string * unit_addr * unit_addr
exception UnknownBehavior of unit_addr * string
exception UnknownGloVersion of string * version

let string_of_error = function
  | CircularDependency ual -> List.fold_left
    (fun s (fn,un) ->
      Printf.sprintf "%s%s#n=%d\n" s fn un
    ) "Circular dependency linking:\n" ual
  | MissingMacro ((fn,un),mn) ->
    Printf.sprintf "%s#n=%d requires macro '%s' which cannot be found.\n" fn un mn
  | MissingSymbol ((fn,un),mn) ->
    Printf.sprintf "%s#n=%d requires symbol '%s' which cannot be found.\n" fn un mn
  | SymbolConflict (ssym,csym,(sfn,sun),(cfn,cun)) ->
    Printf.sprintf "%s#n=%d provides '%s' but exposes '%s' which conflicts with %s#n=%d\n"
      sfn sun ssym csym cfn cun
  | UnknownBehavior ((fn,un),b) ->
    Printf.sprintf "%s#n=%d uses unknown extension behavior '%s'.\n" fn un b
  | UnknownGloVersion (fn,(maj,min,rev)) ->
    Printf.sprintf "%s declares unsupported version %d.%d.%d.\n"
      fn maj min rev
  | exn -> raise exn

(* Prepare the packaged source for string concatenation. *)
let armor meta (url,linkmap,fs_ct) opmac s =
  (* Replace special symbols in line directives to satisfy linkmap *)
  (*required regexp (greetz jwz) instead of macros due to ANGLE bug 183*)
  let intpatt = Re_str.regexp "GLOC_\\([0-9]+\\)" in
  let offset s =
    let fn = int_of_string (Re_str.string_after (Re_str.matched_string s) 5) in
    let anno = try "/* "^url^(List.assoc (string_of_int fn) linkmap)^" */"
    with Not_found -> ""
    in (string_of_int (fn + fs_ct))^anno
  in
  let s = Re_str.global_substitute intpatt offset s in
  let head = match meta with None -> "" | Some meta ->
    let field name (entity,href) = "// "^name^": "^entity^" <"^href^">\n" in
    let authors = List.fold_left
      (fun s link -> s^(field "Author" link))
      "" meta.author in
    let license = match meta.license with None -> ""
      | Some link -> field "License" link in
    let library = match meta.library with None -> ""
      | Some link -> field "Library" link in
    let version = match meta.version with None -> ""
      | Some (v,url) ->
        "// Version: "^(string_of_version v)^" <"^url^">\n" in
    let build = match meta.build with None -> ""
      | Some link -> field "Build" link in
    let (year,(holder,url)) = meta.copyright in
      "// Copyright "^(string_of_int year)^" "^holder^" <"^url^"> "
      ^"All rights reserved.\n"^license^authors^library^version^build
  in let s = head^s in
    (* Undefine local open macros so they do not leak *)
    List.fold_left (fun s mac -> s^"\n#undef "^mac) s opmac

(* Search a glo for a unit satisfying the supplied predicate p. *)
let search p glo =
  let rec loop = function
    | [] -> None
    | (i,u)::r -> if p u then Some i else loop r
  in loop (Array.to_list (Array.mapi (fun i u -> (i,u)) glo.units))

(* Linearly search for an exported macro. *)
let rec satisfy_mac addr macro = function
  | (name,glo)::rest ->
      begin match search (fun u -> List.mem macro u.outmac) glo with
        | Some i -> (name,i)
        | None -> satisfy_mac addr macro rest
      end
  | [] -> raise (MissingMacro (addr, macro))

(* Linearly search for an exported symbol or macro. *)
let rec satisfy_sym addr sym = function
  | (name,glo)::rest ->
      begin match search
        (fun u -> (List.mem sym u.outmac) || (List.mem sym u.outsym)) glo
      with
        | Some i -> (name,i)
        | None -> satisfy_sym addr sym rest
      end
  | [] -> raise (MissingSymbol (addr, sym))

let map_of_list v = List.fold_left (fun m n -> M.add n v m) M.empty

(* Construct the constraint data structure from a glo unit. *)
let tooth addr u =
  { rsym=u.insym; (* to be satisfied *)
    rmac=u.inmac;
    tsym=M.empty; (* locally satisfied *)
    tmac=M.empty;
    bsym=map_of_list addr u.outsym; (* cumulatively satisfies *)
    bmac=map_of_list addr u.outmac;
    addr }

let lookup glo_alist (n,u) = try (List.assoc n glo_alist).units.(u) with _ -> raise (Failure "lookup")
let tooth_of_addr glo_alist addr = tooth addr (lookup glo_alist addr)
let has_addr addr_a ({addr=addr_b}) = addr_a = addr_b
(* Advertize prior units to later units. *)
let mergeb b = function [] -> b
  | {bsym; bmac}::r -> {b with bsym=M.fold M.add b.bsym bsym;
                          bmac=M.fold M.add b.bmac bmac}
(* Satisfy a macro dependency with an already included dependency. *)
let connect_mac b n addr =
  {b with rmac=List.filter (fun m -> not (m=n)) b.rmac; tmac=M.add n addr b.tmac}
(* Satisfy a symbol dependency with an already included dependency. *)
let connect_sym b n addr =
  {b with rsym=List.filter (fun m -> not (m=n)) b.rsym; tsym=M.add n addr b.tsym}
(* Search for an already included macro. *)
let provided_mac n = function [] -> None
  | t::r -> try let addr = M.find n t.bmac in Some addr
    with Not_found -> None
(* Search for an already included symbol. *)
let provided_sym n = function [] -> None
  | t::r -> try let addr = M.find n t.bmac in Some addr
    with Not_found -> try let addr = M.find n t.bsym in Some addr
    with Not_found -> None
(* Given a tooth and a prefix tooth list, find conflicts. *)
let conflicted a = function [] -> None
  | t::r -> let keys = List.map fst ((M.bindings a.bsym)@(M.bindings a.bmac)) in
      begin match List.filter (fun (sym,addr) -> List.mem sym keys)
        ((M.bindings t.bsym)@(M.bindings t.bmac))
      with [] -> None | b::_ -> Some b end

(* Check for symbol conflicts *)
let check_conflicts n tooth (b,t) = match conflicted tooth t with
  | Some (sym, caddr) -> raise (SymbolConflict (n,sym,tooth.addr,caddr))
  | None -> ()

(* Check for circular dependency *)
let check_circdep addr (b,t) =
  if List.exists (has_addr addr) b
  then raise (CircularDependency (List.map (fun {addr} -> addr) b))

(* Build a list of units with internal requirements satisfied. *)
let rec satisfy_zipper glo_alist = function
    (* At the bottom of the zipper, we must be done. *)
  | ([],t) -> ([],t)
    (* Without further needs from below, we must be ready to descend. *)
  | (({rmac=[]; rsym=[]} as b)::r,t) ->
      satisfy_zipper glo_alist (r,(mergeb b t)::t)
    (* Subsequent units require a macro. *)
  | (({rmac=n::_} as b)::r,t) ->
      begin match provided_mac n t with
        | Some addr -> satisfy_zipper glo_alist ((connect_mac b n addr)::r,t)
        | None -> let addr = satisfy_mac b.addr n glo_alist in
            check_circdep addr (b::r,t);
            let tooth = tooth_of_addr glo_alist addr in
              check_conflicts n tooth (b::r,t);
              satisfy_zipper glo_alist(tooth::(connect_mac b n addr)::r,t)
      end
    (* Subsequent units require a symbol. *)
  | (({rmac=[]; rsym=n::_} as b)::r,t) ->
      begin match provided_sym n t with
        | Some addr -> satisfy_zipper glo_alist ((connect_sym b n addr)::r,t)
        | None -> let addr = satisfy_sym b.addr n glo_alist in
            check_circdep addr (b::r,t);
            let tooth = tooth_of_addr glo_alist addr in
              check_conflicts n tooth (b::r,t);
              satisfy_zipper glo_alist (tooth::(connect_sym b n addr)::r,t)
      end

(* Generate a list of unit addresses from a list of required symbols and
   a search list. *)
let sort required glo_alist =
  let addrs = List.fold_left
    (fun al sym -> let addr = satisfy_sym ("[-u "^sym^"]",0) sym glo_alist in
       if List.mem addr al then al else addr::al)
    [] (List.rev required) in
  let z = satisfy_zipper glo_alist
    (List.rev_map (tooth_of_addr glo_alist) addrs,[])
  in List.rev_map (fun tooth -> tooth.addr) (snd z)

(* Generate the shader preamble. *)
let preamble glol =
  (* Precedence for conflicting behaviors *)
  let b_order = ["require"; "warn"; "enable"; "disable"] in
  let rec b_max x y = function [] -> x
    | b::r when b=x -> x
    | b::r when b=y -> y
    | _::r -> b_max x y r
  in
  let ext_merge addr m (ext,b) =
    match (try Some (M.find ext m) with Not_found -> None) with
      | None -> if List.mem b b_order
        then M.add ext b m
        else raise (UnknownBehavior (addr, b))
      | Some pb -> M.add ext (b_max b pb b_order) m
  in
  let ext_decl ext b = "#extension "^ext^" : "^b^"\n" in
  let ext_segment m =
    (try ext_decl "all" (M.find "all" m) with Not_found -> "")
    ^(M.fold (fun ext b s -> if ext="all" then s else s^(ext_decl ext b)) m "")
  in
  let (version,pragmas,exts) = List.fold_left
    (fun (version,pragmas,exts) ((name,i),glo) -> let u = try glo.units.(i) with _ -> raise (Failure "preamble") in
       (begin match u.vdir,version with
          | None,v -> v
          | Some uv,None -> Some uv
          | Some uv,Some pv -> Some (max uv pv)
        end,
        List.fold_left (fun p s -> p^s^"\n") pragmas u.pdir,
        List.fold_left (ext_merge (name,i)) exts u.edir
       )
    ) (None,"",M.empty) glol
  in (match version with Some v -> "#version "^(string_of_int v)^"\n"
        | None -> "")^pragmas^(ext_segment exts)

(* Produce a string representing a valid SL program given a list of required
   symbols and a search list. *)
let link prologue required glo_alist =
  let required = if (List.length required) = 0 then ["main"] else required in
  let support = [|[||];[|true|]|] in
  let () = List.iter
    (fun (name,glo) -> let (maj,min,_) = glo.glo in
       if try not support.(maj).(min) with Invalid_argument _ -> true
       then raise (UnknownGloVersion (name,glo.glo))) glo_alist in
  let glol = List.map
    (fun (name,u) -> ((name,u),List.assoc name glo_alist))
    (sort required glo_alist)
  in fst begin List.fold_left
        begin fun (src,(pname,o)) ((name,u),glo) ->
          let sup = List.fold_left
            (fun sup (is,_) -> max sup (int_of_string is))
            0 glo.linkmap
          in
          let meta = if name=pname then None else glo.meta in
          let u = try glo.units.(u) with _ -> raise (Failure "u") in
          let unit_begin = if name=pname || pname=""
          then "" else "// End: Copyright\n"
          in (src^unit_begin^(armor meta (name, glo.linkmap, o) u.opmac u.source)^"\n",
              (name,o+sup+1))
        end ((preamble glol)^prologue,("",0)) glol
    end

(* Some link-time functions *)

(* Flatten a glom into an association list and remove non-glo elements *)
let flatten prefix glom =
  let rec descend prefix l = function
    | (n,Leaf glo) -> (prefix^n, glo)::l
    | (n,Glom glom) -> List.fold_left
      (fun l p -> descend (prefix^n^"/") l p) l glom
    | (n,Source _) | (n,Other _) -> l
  in match glom with
    | Glom glom -> List.rev
      (List.fold_left
         (fun l p -> descend prefix l p) [] glom)
    | Leaf glo -> [prefix,glo]
    | Source _ | Other _ -> []

(* Nest a glo alist back into a glom *)
let nest glo_alist =
  let rec group prefix prev = function
    | ((_::[],_)::_) as glom -> (glom, List.rev prev)
    | (x::xs,glo)::r when x=prefix -> group prefix ((xs,glo)::prev) r
    | glom -> (glom,List.rev prev)
  in
  let rec nest prev = function
    | [] -> Glom (List.rev prev)
    | (([],glo)::r) -> nest (("", glo)::prev) r
    | ((x::[],glo)::r) -> nest ((x, glo)::prev) r
    | ((x::_,_)::_) as r ->
      let (rest,g) = group x [] r in
      nest ((x, nest [] g)::prev) rest
  in let split s = Re_str.split (Re_str.regexp_string "/") s in
     nest [] (List.map (fun (n,glo) -> (split n, glo)) glo_alist)

File: src/node/glol/lib/glol.js
/**
 * @preserve Copyright 2012 Ashima Arts. All rights reserved.
 * Author: David Sheets
 * Licensed under BSD-3-Clause found at http://www.spdx.org/licenses/BSD-3-Clause
 *
 * @constructor
 */
window['GLOL'] = function() {};

(function() {
  // Let's link some glo!

  // There are a number of exceptions that the algorithm can emit
  /** @constructor */
  var MissingMacro = function(addr,macro) {
    this.name = "MissingMacro"; this.addr = addr; this.macro = macro;
  };
  /** @constructor */
  var MissingSymbol = function(addr,sym) {
    this.name = "MissingSymbol"; this.addr = addr; this.symbol = sym;
  };
  /** @constructor */
  var NotFound = function(key) {
    this.name = "NotFound"; this.key = key;
  };
  /** @constructor */
  var CircularDependency = function(addrl) {
    this.name = "CircularDependency"; this.addrs = addrl;
  };
  /** @constructor */
  var SymbolConflict = function(n,sym,addr,caddr) {
    this.name = "SymbolConflict"; this.sym_a = n; this.sym_b = sym;
    this.addr_a = addr; this.addr_b = caddr;
  };
  /** @constructor */
  var UnknownBehavior = function(addr, b) {
    this.name = "UnknownBehavior"; this.addr = addr; this.behavior = b;
  };
  /** @constructor */
  var UnknownGloVersion = function(path,version) {
    this.name = "UnknownGloVersion"; this.path = path; this.version = version;
  };

  // Prepare the packaged source for concatenation.
  /**
   * @type {function({copyright: Array.<number|Array.<string>>},*,*,*,*) : string }
   */
  function armor(meta,name,linkmap,fs_ct,opmac,s) {
    // Replace special symbols in line directives to satisfy linkmap
    // required regexp (greetz jwz) instead of macros due to ANGLE bug 183
    var intpatt = /GLOC_([0-9]+)/g;
    function f(s,n) {
      var anno = "";
      var fn = parseInt(n,10);
      if (linkmap[n]) {
	anno = "/* "+name+linkmap[n]+" */";
      }
      return (""+(fn+fs_ct)+anno);
    }
    s = s.replace(intpatt,f);

    var head = "";
    if (meta) {
      var authors="",license="",library="",version="",build="";
      function field(name,link) {
	return "// "+name+": "+link[0]+" <"+link[1]+">\n";
      }
      if (meta.author) {
	for (var i=0; i < meta.author.length; i++) {
	  authors += field("Author",meta.author[i]);
	}
      }
      if (meta.license) license = field("License",meta.license);
      if (meta.library) library = field("Library",meta.library);
      if (meta.version) {
	var v = meta.version;
	version = "// Version: "+v[0][0]+"."+v[0][1]+"."+v[0][2];
        version+= " <"+v[1]+">\n";
      }
      if (meta.build) build = field("Build",meta.build);
      head = ("// Copyright "+meta.copyright[0]+" "
	      +meta.copyright[1][0]+" <"+meta.copyright[1][1]+"> "
	      +"All rights reserved.\n"+license+authors+library+version+build);
    }

    // Undefine local open macros so they do not leak
    var undefs = "";
    for (var i = 0; i < opmac.length; i++) {
      undefs += "\n#undef "+opmac[i];
    }
    return head+s+undefs;
  }

  // Search a glo for a unit satisfying the supplied predicate p.
  function search(p, glo) {
    for (var u = 0; u < glo.units.length; u++) {
      var unit = glo.units[u];
      if (p(unit)) return u;
    }
    return null;
  }

  // Linearly search for an exported macro.
  function satisfy_mac(addr, macro, glo_alist) {
    for (var i = 0; i < glo_alist.length; i++) {
      var name = glo_alist[i][0], glo = glo_alist[i][1];
      var u = search(function (unit) { return unit.outmac.indexOf(macro) != -1; },
		     glo);
      if (u != null) return [name, u];
    }
    throw (new MissingMacro(addr,macro));
  }

  // Linearly search for an export symbol.
  function satisfy_sym(addr, sym, glo_alist) {
    for (var i = 0; i < glo_alist.length; i++) {
      var name = glo_alist[i][0], glo = glo_alist[i][1];
      var u = search (function(unit) { return (unit.outmac.indexOf(sym) != -1
					       || unit.outsym.indexOf(sym) != -1); },
		      glo);
      if (u != null) return [name, u];
    }
    throw (new MissingSymbol(addr, sym));
  }

  function map_of_list(v,l) {
    var m = {};
    for (var i = 0; i < l.length; i ++) m[l[i]]=v;
    return m;
  }

  // Construct the constraint data structure from a glo unit.
  /** @type {function(Array.<string|number>,
   *                  {insym: Array.<string>, outsym: Array.<string>,
   *                   inmac: Array.<string>, outmac: Array.<string>,
   *                   vdir: number, pdir: string, edir: Array.<Array.<string>>
   *                   })
   *         : {rsym: Array.<string>, rmac: Array.<string>,
   *            tsym: Object.<string,Array.<string|number>>,
   *            tmac: Object.<string,Array.<string|number>>,
   *            bsym: Object.<string,Array.<string|number>>,
   *            bmac: Object.<string,Array.<string|number>>,
   *            addr: Array.<string|number> }}
   */
  function tooth(addr, u) {
    return {
      "rsym":u.insym, "rmac":u.inmac, // Required: to be satisfied
      "tsym":{}, "tmac":{}, // Top: locally satisfied
      // Bottom: cumulatively satisfies
      "bsym":map_of_list(addr,u.outsym), "bmac":map_of_list(addr,u.outmac),
      "addr":addr
    };
  }

  function assoc(key,alist) {
    for (var i = 0; i < alist.length; i++)
      if (alist[i][0] == key) return alist[i][1];
    throw (new NotFound(key));
  }

  function lookup(glo_alist,addr) {
    return assoc(addr[0],glo_alist).units[addr[1]];
  }

  function tooth_of_addr(glo_alist,addr) {
    return tooth(addr,lookup(glo_alist,addr));
  }

  function has_addr(addr,tooth) {
    return tooth.addr[0] == addr[0] && tooth.addr[1] == addr[1];
  }

  // Advertize prior units to later units.
  function mergeb(b,top) {
    if (top.length == 0) return b;
    else {
      var zb = top[0];
      for (var k in zb.bsym) b.bsym[k] = zb.bsym[k];
      for (var k in zb.bmac) b.bmac[k] = zb.bmac[k];
      return b;
    }
  }

  // Satisfy a macro dependency with an already included dependency.
  function connect_mac(b,n,addr) {
    b.rmac=b.rmac.filter(function(m){ return m!=n; });
    b.tmac[n] = addr;
    return b;
  }

  // Satisfy a symbol dependency with an already included dependency.
  function connect_sym(b,n,addr) {
    b.rsym=b.rsym.filter(function(m){ return m!=n; });
    b.tsym[n] = addr;
    return b;
  }

  // Search for an already included macro.
  function provided_mac(n,top) {
    if (top.length == 0) return null;
    else if (n in top[0].bmac) return top[0].bmac[n];
    else return null;
  }

  // Search for an already included symbol.
  function provided_sym(n,top) {
    if (top.length == 0) return null;
    else if (n in top[0].bmac) return top[0].bmac[n];
    else if (n in top[0].bsym) return top[0].bsym[n];
    else return null;
  }

  // Given a tooth and a prefix tooth list, find conflicts.
  function conflicted(tooth,top) {
    if (top.length == 0) return null;
    else {
      for (var key in tooth.bsym) {
	if (key in top[0].bsym) return [key, top[0].bsym[key]];
	if (key in top[0].bmac) return [key, top[0].bmac[key]];
      }
      for (var key in tooth.bmac) {
	if (key in top[0].bsym) return [key, top[0].bsym[key]];
	if (key in top[0].bmac) return [key, top[0].bmac[key]];
      }
    }
    return null;
  }

  // Check for symbol conflicts
  function check_conflicts(n,tooth,zipper) {
    var conflict = conflicted(tooth,zipper[1]);
    if (conflict != null) {
      throw (new SymbolConflict(n,conflict[0],tooth.addr,conflict[1]));
    }
  }

  // Check for circular dependency
  function check_circdep(addr,zipper) {
    if (zipper[0].some(function (tooth){ return has_addr(addr,tooth); })) {
      throw (new CircularDependency(
	       zipper[0].map(function (t) { return t.addr; })));
    }
  }

  // Build a list of units with internal requirements satisfied.
  // TODO: labelled loop + continue to avoid tail calls in TCO-less JS
  function satisfy_zipper(glo_alist,zipper) {
    // At the bottom of the zipper, we must be done.
    if (zipper[0].length == 0) return zipper;
    // Subsequent units require a macro.
    else if (zipper[0][0].rmac.length != 0) {
      var b = zipper[0][0], n = b.rmac[0];
      var addr = provided_mac(n,zipper[1]);
      if (addr != null) {
	connect_mac(b,n,addr);
	return satisfy_zipper(glo_alist,zipper);
      }
      addr = satisfy_mac(b.addr,n,glo_alist);
      check_circdep(addr,zipper);
      var tooth = tooth_of_addr(glo_alist,addr);
      check_conflicts(n,tooth,zipper);
      connect_mac(b,n,addr);
      zipper[0].unshift(tooth);
      return satisfy_zipper(glo_alist,zipper);
    } else // Subsequent units require a symbol.
    if (zipper[0][0].rsym.length != 0) {
      var b = zipper[0][0], n = b.rsym[0];
      var addr = provided_sym(n,zipper[1]);
      if (addr != null) {
	connect_sym(b,n,addr);
	return satisfy_zipper(glo_alist,zipper);
      }
      addr = satisfy_sym(b.addr,n,glo_alist);
      check_circdep(addr,zipper);
      var tooth = tooth_of_addr(glo_alist,addr);
      check_conflicts(n,tooth,zipper);
      connect_sym(b,n,addr);
      zipper[0].unshift(tooth);
      return satisfy_zipper(glo_alist,zipper);
    } else { // Without further needs from below, we are ready to descend.
      var b = zipper[0].shift();
      zipper[1].unshift(mergeb(b,zipper[1]));
      return satisfy_zipper(glo_alist,zipper);
    }
  }

  // Generate a list of unit addresses from a list of required symbols and
  // a search list
  function sort(reqsym,glo_alist) {
    var addrs = reqsym.reduce(function (al,sym) {
      var addr = satisfy_sym(["[-u "+sym+"]",0],sym,glo_alist);
      if (al.every(function (a) { return a[0]!=addr[0] || a[1]!=addr[1]; })) {
	al.push(addr);
      }
      return al;
    },[]);
    var zipper = [addrs.map(function (addr) {
			      return tooth_of_addr(glo_alist,addr);
			    }),[]];
    zipper = satisfy_zipper(glo_alist,zipper);
    return zipper[1].reduce(function (al,tooth) {
			      al.unshift(tooth.addr);
			      return al;
			    },[]);
  }

  // Generate the shader preamble.
  function preamble(glol) {
    // Precedence for conflicting behaviors
    var b_order = ["require","warn","enable","disable"];
    function b_max(x,y) {
      return b_order[Math.min(b_order.indexOf(x),b_order.indexOf(y))];
    }
    function ext_merge(addr,m,ext) {
      if (ext[0] in m) {
	m[ext[0]] = b_max(ext[1],m[ext[0]]); return m;
      } else if (b_order.indexOf(ext[1]) >= 0) {
	m[ext[0]] = ext[1]; return m;
      } else throw (new UnknownBehavior(addr,ext[1]));
    }
    function ext_decl(ext,b) { return "#extension "+ext+" : "+b+"\n"; }
    function ext_segment(m) {
      var exts = "";
      for (var e in m) if (e!="all") exts += ext_decl(e,m[e]);
      return ((("all" in m)?ext_decl("all",m.all):"")+exts);
    }
    var preamble = glol.reduce(function(pre,gap) {
      var u = gap[1].units[gap[0][1]];
      return [u.vdir==null?pre[0]
	      :(pre[0]==null?u.vdir
		:Math.max(u.vdir,pre[0])),
	      u.pdir.reduce(function(p,s) { return p+s+"\n"; },pre[1]),
	      u.edir.reduce(function(m,ext) {
			      return ext_merge(gap[0],m,ext);
			    },pre[2])
	     ];
      },[null,"",{}]);
    return (((preamble[0]!=null)?("#version "+preamble[0]+"\n"):"")
	    +preamble[1]+(ext_segment(preamble[2])));
  }

  function flatten(prefix,glom) {
    return glom.reduce(function(glo_alist,ngp) {
      if (Array.isArray(ngp[1]))
	return glo_alist.concat(flatten(prefix+ngp[0]+"/",ngp[1]));
      else return glo_alist.concat([[prefix+ngp[0],ngp[1]]]);
    },[]);
  }

  function filter(alist) {
    var l = [];
    for (var i = 0; i < alist.length; i++) {
      var glo = alist[i][1];
      if (glo.glo) {
	if (glo.glo[0] != 1 || glo.glo[1] != 0) {
	  throw (new UnknownGloVersion(alist[i][0],glo.glo));
	}
	l.push(alist[i]);
      }
    }
    return l;
  }

  // Zeros may be omitted for transport; add them back.
  function add_zero(o,f,z) { if (!(f in o)) { o[f] = z; } return o; }

  function add_zeros(alist) {
    return alist.map(function (ngp) {
      var name = ngp[0], glo = ngp[1];
      add_zero(glo,"units",[]);
      add_zero(glo,"linkmap",{});
      glo.units = glo.units.map(function (u) {
	add_zero(u,"pdir",[]); add_zero(u,"edir",[]);
	add_zero(u,"insym",[]);	add_zero(u,"outsym",[]);
	add_zero(u,"inmac",[]);	add_zero(u,"opmac",[]);	add_zero(u,"outmac",[]);
	return u;
      });
      return [name,glo];
    });
  }

  // Produce a string representing a valid SL program given a list of required
  // symbols and a search list.
  /** @type {function(string,Array.<string>,*) : string } */
  window['GLOL'].prototype.link = function(prologue,reqsym,glom) {
    var reqsym_ = (reqsym.length==0) ? ["main"] : reqsym.reverse();
    var glo_alist = add_zeros(filter(flatten("",glom)));
    var glol = (sort(reqsym_,glo_alist)).map(function (addr) {
      return [addr,assoc(addr[0],glo_alist)];
    });
    return glol.reduce(function (acc,agp) {
      var sup = 0, glo = agp[1], name=agp[0][0], src=acc[0],
	  pname=acc[1][0], o=acc[1][1];
      for (var k in glo.linkmap) { sup = Math.max(sup,parseInt(k,10)); }
      var u = glo.units[agp[0][1]];
      var unit_begin = (name==pname||pname=="")?"":"// End: Copyright\n";
      return [(src+unit_begin+armor(name==pname?null:glo.meta,name,
				    glo.linkmap,o,u.opmac,u.source)+"\n"),
	      [name,o+sup+1]];
    }, [preamble(glol)+prologue,["",0]])[0];
  };
})();
